packages/python-packages/apiview-copilot/evals/run.py

```diff
@@ -5,79 +5,145 @@
-        """Compare violations based on both line numbers and rule IDs."""
+    def _get_comment_matches(self, expected: dict[str, Any], actual: dict[str, Any]) -> Tuple[Set, Set, Set]:
+        """Compare comments based on both line numbers and rule IDs."""
         exact_matches = set()
         rule_matches_wrong_line = set()
-        line_matches_wrong_rule = set()
 
-        violations_left = copy.deepcopy(actual["violations"])
-        for expected_violation in expected["violations"]:
-            e_line = expected_violation["line_no"]
-            e_rules = frozenset(expected_violation["rule_ids"])
+        comments_left = copy.deepcopy(actual["comments"])
+        for expected_comment in expected["comments"]:
+            e_line = expected_comment["line_no"]
+            e_rules = frozenset(expected_comment["rule_ids"])
 
-            for actual_violation in violations_left:
-                a_line = actual_violation["line_no"]
-                a_rules = frozenset(actual_violation["rule_ids"])
+            for actual_comment in comments_left:
+                a_line = actual_comment["line_no"]
+                a_rules = frozenset(actual_comment["rule_ids"])
 
                 rule_match = any(rule for rule in a_rules if rule in e_rules)
                 if e_line == a_line and rule_match:
                     exact_matches.add((e_line, tuple(sorted(e_rules))))
-                    # Remove the matched actual violation to avoid double counting
-                    violations_left.remove(actual_violation)
+                    # Remove the matched actual comment to avoid double counting
+                    comments_left.remove(actual_comment)
                     break
                 if rule_match:
                     if abs(e_line - a_line) <= 5:
                         # If the line numbers are close, consider it a match
                         rule_matches_wrong_line.add((tuple(sorted(e_rules)), e_line, a_line))
-                elif e_line == a_line:
-                    line_matches_wrong_rule.add((e_line, tuple(sorted(e_rules)), tuple(sorted(a_rules))))
-
-        return exact_matches, rule_matches_wrong_line, line_matches_wrong_rule
+                        comments_left.remove(actual_comment)
+                        break
+
+        return exact_matches, rule_matches_wrong_line, comments_left
+
+    def _evaluate_generic_comments(self, query: str, language: str, generic_comments: list[dict[str, Any]]) -> None:
+        """Evaluate generic comments. If they are invalid, they count as false positives and receive penalty."""
+
+        filter_path = pathlib.Path(__file__).parent.parent / "metadata" / language / "filter.yaml"
+        with open(filter_path, "r") as f:
+            filter_data = yaml.safe_load(f)
+            exceptions = filter_data["exceptions"].strip().split("\n")
+            exceptions = [e.split(". ", 1)[1] for e in exceptions]
+
+        client = openai.AzureOpenAI()
+        for comment in generic_comments:
```

Comment:
Since this is a query per generic comment, I'd recommend running these in parallel. Assuming o4-mini takes as long as o3-mini, running these in series would really drag out the run times.

Reply:
gpt-4o is used as the model judge for all evals (including generic comments), and it's much faster than the reasoning models. I didn't sense a difference in run times by adding this. Overall, I do need to investigate how we can speed up evals, but I'm going to work on that in another PR (tracked in https://github.com/Azure/azure-sdk-tools/issues/10445)